fv3core/stencils/divergence_damping.py

```diff
@@ -149,90 +149,138 @@ class DivergenceDamping:
+        assert grid_type < 3, "Not implemented, grid_type>=3, specifically smag_corner"
+        # TODO: make dddmp a compile-time external, instead of runtime scalar
+        self._dddmp = dddmp
+        # TODO: make da_min_c a compile-time external, instead of runtime scalar
+        self._da_min_c = da_min_c
+        self._grid_type = grid_type
         self._nord_column = nord_col
         self._d2_bg_column = d2_bg
+        self._divg_u = divg_u
+        self._divg_v = divg_v
+        self._rarea_c = rarea_c
+        self._sin_sg1 = sin_sg1
+        self._sin_sg2 = sin_sg2
+        self._sin_sg3 = sin_sg3
+        self._sin_sg4 = sin_sg4
+        self._cosa_u = cosa_u
+        self._cosa_v = cosa_v
+        self._sina_u = sina_u
+        self._sina_v = sina_v
+        self._dxc = dxc
+        self._dyc = dyc
         self._nonzero_nord_k = 0
-        self._nonzero_nord = int(namelist.nord)
+        self._nonzero_nord = int(nord)
         for k in range(len(self._nord_column)):
             if self._nord_column[k] > 0:
                 self._nonzero_nord_k = k
                 self._nonzero_nord = int(self._nord_column[k])
                 break
-        if self.grid.stretched_grid:
-            self._dd8 = self.grid.da_min * self._d4_bg ** (self._nonzero_nord + 1)
+        if stretched_grid:
+            self._dd8 = da_min * d4_bg ** (self._nonzero_nord + 1)
         else:
-            self._dd8 = (self.grid.da_min_c * self._d4_bg) ** (self._nonzero_nord + 1)
+            self._dd8 = (da_min_c * d4_bg) ** (self._nonzero_nord + 1)
+        # TODO: make stretched_grid a compile-time external, instead of runtime scalar
+        self._stretched_grid = stretched_grid
         kstart = self._nonzero_nord_k
         low_kstart = 0
-        nk = self.grid.npz - kstart
+        nk = self._idx.ke - kstart
         low_nk = self._nonzero_nord_k
         self.a2b_ord4 = AGrid2BGridFourthOrder(
-            self._grid_type, kstart, nk, replace=False
+            grid_indexing,
+            agrid1,
+            agrid2,
+            bgrid1,
+            bgrid2,
+            dxa,
+            dya,
+            edge_n,
+            edge_s,
+            edge_e,
+            edge_w,
+            self._grid_type,
+            kstart,
+            nk,
+            replace=False,
         )
 
-        start_points = axis_offsets(
-            self.grid,
-            (self.grid.is_ - 1, self.grid.js, low_kstart),
-            (self.grid.nic + 2, self.grid.njc + 1, low_nk),
-        )
-        self._ptc_computation = FrozenStencil(
+        self._ptc_computation = self._get_stencil(
             ptc_computation,
-            origin=(self.grid.is_ - 1, self.grid.js, low_kstart),
-            domain=(self.grid.nic + 2, self.grid.njc + 1, low_nk),
-            externals=start_points,
+            dims=[X_DIM, Y_INTERFACE_DIM],
+            halos=(1, 0),
+            kstart=low_kstart,
+            nk=low_nk,
         )
```

Comment:
This looks a little ugly in the changelog, but the final code here looks really nice, I promise!

Reply:
Yes I agree there's some iteration to do here to merge this complexity into `GridIndexing`.

Reply:
This is refactored now.